### ISSUE TITLE
Container logs with timestamp are getting trailing zeroes trimmed

### ISSUE THREAD

**Issue by smarterclayton** [kubernetes/kubernetes#72292]
Container logs with timestamp=true display a ragged left edge on 1.11.2 with cri-o:

```
2018-12-22T00:34:33.084862626Z Pushed 5/6 layers, 83% complete
2018-12-22T00:35:03.196483456Z Pushed 5/6 layers, 87% complete
2018-12-22T00:35:33.203214994Z Pushed 5/6 layers, 89% complete
2018-12-22T00:36:03.271229927Z Pushed 5/6 layers, 92% complete
2018-12-22T00:36:33.34432357Z Pushed 5/6 layers, 96% complete
2018-12-22T00:37:03.444918533Z Pushed 5/6 layers, 98% complete
2018-12-22T00:37:36.90513849Z Pushed 6/6 layers, 100% complete
2018-12-22T00:37:41.364784926Z Push successful
```

It looks like when the nano section is written we're dropping the right most precision character (possibly because something is assuming it's a trailing zero, like a format).  Another section shows two missing chars:

```
2018-12-22T00:33:13.638865611Z Spawning worker 0 with 2 pkgs
2018-12-22T00:33:13.6388949Z Spawning worker 1 with 2 pkgs
2018-12-22T00:33:13.638899415Z Spawning worker 2 with 2 pkgs
```

which makes it seem likely that something is stripping trailing zeros off.  We should preserve them or otherwise people will go mad staring at the ragged logs.

@runcom @mrunalp since I'm not sure if this is crio, CRI, or Kubelet

/kind bug

**Comment by vikaschoudhary16** [comment]
@smarterclayton issue is from the time format parsing logic in the CRI. [It uses time-format `RFC3339Nano`](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kuberuntime/logs/logs.go#L51). `time` package then while parsing for this format hits [this if condition](https://github.com/golang/go/blob/master/src/time/format.go#L262) which decides it to be of type `stdFracSecond9` which specifically [says](https://github.com/golang/go/blob/master/src/time/format.go#L122) that trailing zeroes will be omitted for this format.

One way to keep trailing zeros is to use [StampNano( "Jan _2 15:04:05.000000000") format](https://github.com/golang/go/blob/master/src/time/format.go#L85) in the `kuberuntime`, but then this format does not look that detailed and missing year information and changing format in kuberuntime will also require changes in the [parseCRIlog()](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kuberuntime/logs/logs.go#L128) in the kuberuntime.

OR another approach could be to add a new time format in GO's time package, which will have all the details + not omit trailing zeros.

looks like so much pain :)

**Comment by smarterclayton** [comment]
Yeah, but this is ugly as all hell.  I don’t actually care about the
trailing zeros but the alignment of the message column is more important.
The format is pretty unreadable.



On Dec 22, 2018, at 4:18 AM, Vikas Choudhary (vikasc) <
notifications@github.com> wrote:

@smarterclayton <https://github.com/smarterclayton> issue is from the time
format parsing logic in the CRI. It uses time-format RFC3339Nano
<https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kuberuntime/logs/logs.go#L51>.
time package then while parsing for this format hits this if condition
<https://github.com/golang/go/blob/master/src/time/format.go#L262> which
decides it to be of type stdFracSecond9 which specifically says
<https://github.com/golang/go/blob/master/src/time/format.go#L122> that
trailing zeroes will be omitted for this format.

One way to keep trailing zeros is to use StampNano( "Jan _2
15:04:05.000000000") format
<https://github.com/golang/go/blob/master/src/time/format.go#L85> in the
kuberuntime, but then this format does not look that detailed and missing
year information and changing format in kuberuntime will also require
changes in the parseCRIlog()
<https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kuberuntime/logs/logs.go#L128>
in the kuberuntime.

OR another approach could be to add a new time format in GO's time package,
which will have all the details + not omit trailing zeros.

looks like so much pain :)

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/kubernetes/kubernetes/issues/72292#issuecomment-449557227>,
or mute the thread

**Comment by Nuru** [comment]
@smarterclayton @vikaschoudhary16 I have proposed a fix for this in #91186. I do not think it is a hairy or difficult fix. Please vote for and otherwise support it if you agree.
